### PR TITLE
Fix inconsistent client lerp and account for blend factor on server

### PR DESCRIPTION
**Pull request by AMZN-puvvadar**
Client proxies can experience some jitter from lerping when completely still. This is caused by blend factor varying based on lag, breakpoints or really anything that affects time. Updating the previous transform ensures we're moving closer to the target vs potentially backwards. This does fundamentally change the rate of interpolation but it visually looked reasonable to me.

Signed-off-by: puvvadar <puvvadar@amazon.com>

### CODE REVIEW

**Review by SergeyAMZN — APPROVED** [review]

**Review by kberg-amzn — COMMENTED** [review]
The one major constraint this breaks is we need a way to reconstruct the blended transform value on the server using the blend factor (see the valve article on backward reconciliation for a more detailed explanation of why).. using this approach we’d have to transmit a sequence of blend factors that the server would have to iteratively apply to reach the same values.. otherwise I agree this would be a far better way to blend between two transforms

**Comment by AMZN-puvvadar** [comment]
Attempting a different approach here. Fundamentally the problem is that when MPSystemComponent updates m_renderBlendFactor, NetworkTransformComponents do not account for this. This results in the previous and target used for lerping to remain the same while the host frame ID has advanced to the point that the transform should be at its target.

To handle I've added host frame tracking to NetworkTransformComponent. OnPreRender uses this to check if host frame ID has exceeded the transform's target. If so, it'll adjust previous to target to effectively finalize the lerp. Motion remains smooth while jittering has stopped.

As part of this I also confirmed that non-character client proxies were also jittering though in extremely small amounts.

**Review by kberg-amzn — COMMENTED** [review]
This is a good start! Some ideas.. can you merge the blend factor and host time ms in the input structure? seems like there's some duplicated data there..

Also, at least these two components:
- NetworkTransformComponent
- NetworkHitVolumesComponent

During the sync time callback, we want to check that blend factor and use it to sample the correct rewindable values for transforms and hit volumes and duplicate the blending logic that a client has performed during pre-render. So as an example:

Client pre-render, blends game time forward
- we compute a blended transform interpolating to latest network state
- we perform an input processing event based on the current pre-rendered data presented to screen
- we send that input to the server
- server now receives input rpc
- server has the base host frame id, and will forcibly rewind to that host frame id and game time
- server invokes rewind sync event
- network transform component on server hooks rewind sync event
- network transform component samples the blend time and understands it needs to look at both hostframeid and hostframeid - 1 transforms from the rewind buffer
- network transform component fetches those two transforms and blends them using the blend factor to recreate the transform that was presented during client input processing

**Comment by AMZN-puvvadar** [comment]
Merging the input structure is something I'm going to look at for sure. I'm not yet 100% confident if that'll provide enough detail so it'll probably be the last thing I do here. I need to dig into host frame history a bit more to handle some of the blending considerations. I'll likely make use of the existing SyncRewind machinery here to get the state proper.

**Comment by kberg-amzn** [comment]
> Merging the input structure is something I'm going to look at for sure. I'm not yet 100% confident if that'll provide enough detail so it'll probably be the last thing I do here. I need to dig into host frame history a bit more to handle some of the blending considerations. I'll likely make use of the existing SyncRewind machinery here to get the state proper.

Makes sense.. you're almost certainly going to have to punch a new interface into rewindable object. We don't want this blending to be universal though, especially for some values blending may not make any sense at all, enumerations, etc.. really we only want to perform this blend reconstruction for values that affect the positions of entities within the world, so pretty much limited to the network transform component and network hit volumes (possibly the network character to properly sync the pxController base position if that affects locally predicted input processing for some reason).

**Comment by AMZN-puvvadar** [comment]
Added mechanisms to enable interpolation during rewind

- GetPrevious to RewindableObject, this enables grabbing the value before whatever the current time points at
- Updating ScopedAlterTime to take a blend factor. ScopedAlterTime is our main mechanism to rewind time so injecting blend factor here made sense

I have this tested in MPSample where NetworkRigidBody and AnimatedHitVolumes make use of this. A change in that repo will be up in a bit.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl [hunk 1]
```diff
@@ -66,6 +66,12 @@ namespace Multiplayer
         return GetValueForTime(GetCurrentTimeForProperty());
     }
 
+    template <typename BASE_TYPE, AZStd::size_t REWIND_SIZE>
+    inline const BASE_TYPE& RewindableObject<BASE_TYPE, REWIND_SIZE>::GetPrevious() const
+    {
+        return GetValueForTime(GetCurrentTimeForProperty() - HostFrameId(1));
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl:72]: Ooph, careful.. put that -1 inside the GetCurrentTimeForProperty() instead because this will do really bad things on the autonomous client. For more clarity, when we perform the scoped rewind on the server, we *don't* rewind the predictable properties of the autonomous player because we're just about to tick them forward. Basically, 'Don't rewind the shooter'.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/INetworkTime.h [hunk 2]
```diff
@@ -60,6 +64,10 @@ namespace Multiplayer
         //! @param rewindConnectionId the rewinding ConnectionId 
         virtual void AlterTime(HostFrameId frameId, AZ::TimeMs timeMs, AzNetworking::ConnectionId rewindConnectionId) = 0;
 
+        //! Alters the current Host blend factor. Used to drive interpolation in rewound states.
+		//! @param blendFactor the blend factor to use
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/INetworkTime.h:68]: Indentation looks a bit wonky.. also why not add the blend factor as a parameter to AlterTime itself?

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayer.h [hunk 3]
```diff
@@ -182,18 +187,20 @@ namespace Multiplayer
...
             m_previousHostTimeMs = time->GetHostTimeMs();
             m_previousRewindConnectionId = time->GetRewindingConnectionId();
             time->AlterTime(frameId, timeMs, connectionId);
+            time->AlterBlendFactor(blendFactor);
         }
         inline ~ScopedAlterTime()
         {
             INetworkTime* time = GetNetworkTime();
             time->AlterTime(m_previousHostFrameId, m_previousHostTimeMs, m_previousRewindConnectionId);
+            time->AlterBlendFactor(DefaultBlendFactor);
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayer.h:203]: Don't we want this to do the same snap off current blend factor and restore blend factor? As an example, nested scoped alter times are something that happens during backward reconciliation of projectiles.

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp [hunk 4]
```diff
@@ -155,9 +155,13 @@ namespace Multiplayer
             // Discard move input events, client may be speed hacking
             if (m_clientBankedTime < sv_MaxBankTimeWindowSec)
             {
+                // Client blends from previous frame to target so here we subtract blend factor to get to that state
+                const float blendFactor = AZStd::max(0.f, input.GetHostBlendFactor());
+                const float adjustedBlendFactor = std::pow(0.2f, blendFactor);
+                const AZ::TimeMs blendMs = AZ::TimeMs(static_cast<float>(static_cast<AZ::TimeMs>(cl_InputRateMs)) * adjustedBlendFactor);
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:161]: This is really nice!

Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:161]: Just a thought, we should factor out the blend function even though it's just a single 'pow(0.2f, value)', in case we find it's not adequate or buggy or we just want to try out different blend functions. Also do you want to clamp that blend factor between 0 and 1?

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp [hunk 5]
```diff
@@ -311,7 +315,7 @@ namespace Multiplayer
             ++ModifyLastInputId();
             input.SetClientInputId(GetLastInputId());
 
-            ScopedAlterTime scopedTime(input.GetHostFrameId(), input.GetHostTimeMs(), invokingConnection->GetConnectionId());
+            ScopedAlterTime scopedTime(input.GetHostFrameId(), input.GetHostTimeMs(), DefaultBlendFactor, invokingConnection->GetConnectionId());
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:318]: Perform the same time blending on this rewind, it's contained inside your input.

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp [hunk 6]
```diff
@@ -391,7 +395,7 @@ namespace Multiplayer
         {
             // Reprocess the input for this frame
             NetworkInput& input = m_inputHistory[replayIndex];
-            ScopedAlterTime scopedTime(input.GetHostFrameId(), input.GetHostTimeMs(), invokingConnection->GetConnectionId());
+            ScopedAlterTime scopedTime(input.GetHostFrameId(), input.GetHostTimeMs(), DefaultBlendFactor, invokingConnection->GetConnectionId());
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:398]: Same, we want the same blending logic for time that you have above here too. Even though this is on a client correct, when we replay these inputs if one of those inputs relied on some kind of blended state (think two moving players running into each other for example), we need that blended state to be correct on input replay as well.

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp [hunk 7]
```diff
@@ -573,7 +578,7 @@ namespace Multiplayer
 
             NetworkInput& input = m_lastInputReceived[0];
             {
-                ScopedAlterTime scopedTime(input.GetHostFrameId(), input.GetHostTimeMs(), AzNetworking::InvalidConnectionId);
+                ScopedAlterTime scopedTime(input.GetHostFrameId(), input.GetHostTimeMs(), DefaultBlendFactor, AzNetworking::InvalidConnectionId);
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:581]: This is okay with the default here. The client is doomed to correct in any case, and this is happening because the client has starved the server of input packets.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/INetworkTime.h (hunk 2 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/INetworkTime.h:68]: That's what I get for adding comments in my diff tool. I did this separately to make AlterTime a more intentional action rather than a part of AlterTime. For example, when we're adjusting the Client's time based on entity update receipt, we don't want a blend factor. That case was the biggest concern in my mind.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl (hunk 1 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl:72]: That would just subtract a millisecond vs getting us the previous value wouldn't it? I'll need to look at that player case. Granted this also needs more testing. I'm hoping to pull your changes to MPSample in first though.

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp (hunk 4 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:161]: I've had that thought on factoring out the blend function. I was going to wait to grab your changes to blending and then make a pass. I'll want to do that clamp once your changes are in since it changes how we represent blend factor in general. Right now, values above 1 are valid but rare (the pow does the clamp). Your change to send the end value will actually factor a lot of the spots I have to calculate the blend which'll be nice.

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp (hunk 5 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:318]: Okay, I haven't dug too much into the migration code so wasn't sure. I'll update it.

File: Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp (hunk 6 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:398]: That seems reasonable, I'll need to check that input on the client actually has blend factor populated.

File: Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayer.h (hunk 3 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayer.h:203]: Ah okay, I hadn't realized we were nesting. I had thought it was more of an iterative thing. I'll check through the reconciliation code and make updates.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl (hunk 1 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl:72]: Did a more in depth dive here and this should be fine. This will just grab the value for one frame behind perceived current. It's isolated to sync rewind usages.

**Comment by AMZN-puvvadar** [comment]
Latest changes integrated blending formula adjustments from @kberg-amzn and fixes some incorrect math on my part when determining blended time when the server processes input.

I did some breakpoint based testing to verify values. We'll need weapons in place to eye test but I think any resultant changes there can come in another PR.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl (hunk 1 above)
Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl:72]: Nope, you've missed something _really_ critical..  That GetValueForTime call eventually drills down into this function:
```
    HostFrameId NetworkTime::GetHostFrameIdForRewindingConnection(AzNetworking::ConnectionId rewindConnectionId) const
    {
        return (IsTimeRewound() && (rewindConnectionId == m_rewindingConnectionId)) ? m_unalteredFrameId : m_hostFrameId;
    }
```

Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl:72]: Actually.. on further thought, maybe we could just put an assert inside this new function.. if anyone calls GetPrevious on a predictive property for the input processing player; we could just make that an error condition.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.h [hunk 8]
```diff
@@ -155,6 +156,7 @@ namespace Multiplayer
         HostFrameId m_lastReplicatedHostFrameId = HostFrameId(0);
 
         double m_serverSendAccumulator = 0.0;
+		float m_renderBlendFactor = 0.0f;
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.h:159]: indentation bites again? :D

File: Gems/Multiplayer/Code/Source/NetworkTime/NetworkTime.cpp [hunk 9]
```diff
@@ -95,6 +105,7 @@ namespace Multiplayer
 
                     if (networkTransform != nullptr)
                     {
+                        // We're not presently factoring in interpolated position here
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/NetworkTime/NetworkTime.cpp:108]: Good point :\ luckily the error here should be much smaller than velocity * rtt, but we still might want to extrude our intersect volumes a bit to compensate.

**Review by kberg-amzn — APPROVED** [review]

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl (hunk 1 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl:72]: Right, that part was intentional since it's how Get operates. GetPrevious is always the value in history one frame behind Get, even in rewind contexts. That being said, an assert makes sense since there are some contexts where it wouldn't make sense to call. It's really only intended for SyncRewind after all.

File: Gems/Multiplayer/Code/Source/NetworkTime/NetworkTime.cpp (hunk 9 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Source/NetworkTime/NetworkTime.cpp:108]: It's something I plan to test more once weapons are in and we increase the density of the test level. It's an awkward spot to access data to interpolate so I figured let's see what the specific edge cases are before we settle on handling.

File: Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl (hunk 1 above)
Comment (AMZN-puvvadar) [reply Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.inl:72]: I updated the function description for now to clarify the usage. I'll revisit the Assert once I can test edge cases more thoroughly with weapons in place.
